File: apps/api/src/services/tradingDivision.ts
export type TradingDivisionSummary = {
  connected: boolean;
  health: 'healthy' | 'degraded' | 'disconnected';
  updated_at: string;
  certifications: { certified: number; probation: number; rejected: number };
  reconciliation: { clean: boolean | null; issues: number; last_run_at: string | null };
  executions: { recent: number; accepted: number; blocked: number; last_event_at: string | null };
  metrics: Record<string, number>;
  reason?: string;
};

const disconnected = (reason: string): TradingDivisionSummary => ({
  connected: false,
  health: 'disconnected',
  updated_at: new Date().toISOString(),
  certifications: { certified: 0, probation: 0, rejected: 0 },
  reconciliation: { clean: null, issues: 0, last_run_at: null },
  executions: { recent: 0, accepted: 0, blocked: 0, last_event_at: null },
  metrics: {},
  reason,
});

export async function getTradingDivisionSummary(): Promise<TradingDivisionSummary> {
  const endpoint = process.env.TRADING_TELEMETRY_URL || '';
  if (!endpoint) return disconnected('TRADING_TELEMETRY_URL_NOT_CONFIGURED');

  try {
    const response = await fetch(endpoint, {
      headers: { Accept: 'application/json' },
      signal: AbortSignal.timeout(8000),
    });
    if (!response.ok) return disconnected(`TRADING_TELEMETRY_HTTP_${response.status}`);
    const payload = await response.json() as Partial<TradingDivisionSummary>;
    return {
      connected: true,
      health: payload.health === 'degraded' ? 'degraded' : 'healthy',
      updated_at: payload.updated_at || new Date().toISOString(),
      certifications: payload.certifications || { certified: 0, probation: 0, rejected: 0 },
      reconciliation: payload.reconciliation || { clean: null, issues: 0, last_run_at: null },
      executions: payload.executions || { recent: 0, accepted: 0, blocked: 0, last_event_at: null },
      metrics: payload.metrics || {},
    };
  } catch (error) {
    return disconnected(error instanceof Error ? error.message : 'TRADING_TELEMETRY_UNAVAILABLE');
  }
}

File: ai_company/trading_division.py
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

STATE_PATH = Path(os.getenv("AI_COMPANY_TRADING_STATE", ".ai-company-trading.json"))


def _empty(reason: str = "TRADING_TELEMETRY_NOT_INGESTED") -> dict[str, Any]:
    return {
        "connected": False,
        "health": "disconnected",
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "certifications": {"certified": 0, "probation": 0, "rejected": 0},
        "reconciliation": {"clean": None, "issues": 0, "last_run_at": None},
        "executions": {"recent": 0, "accepted": 0, "blocked": 0, "last_event_at": None},
        "metrics": {},
        "reason": reason,
    }


def load_trading_division() -> dict[str, Any]:
    if not STATE_PATH.exists():
        return _empty()
    try:
        payload = json.loads(STATE_PATH.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return _empty("TRADING_TELEMETRY_INVALID")

    if not isinstance(payload, dict):
        return _empty("TRADING_TELEMETRY_INVALID")

    payload.setdefault("connected", True)
    payload.setdefault("health", "healthy")
    payload.setdefault("updated_at", datetime.now(timezone.utc).isoformat())
    payload.setdefault("certifications", {"certified": 0, "probation": 0, "rejected": 0})
    payload.setdefault("reconciliation", {"clean": None, "issues": 0, "last_run_at": None})
    payload.setdefault("executions", {"recent": 0, "accepted": 0, "blocked": 0, "last_event_at": None})
    payload.setdefault("metrics", {})
    return payload


def save_trading_division(payload: dict[str, Any]) -> dict[str, Any]:
    STATE_PATH.write_text(json.dumps(payload, indent=2, default=str), encoding="utf-8")
    return load_trading_division()
